### ISSUE TITLE
Creating context fails with "glxSwapIntervalMESA failed"

### ISSUE THREAD

**Issue by rocurley** [ggez/ggez#294]
Whenever ggez tries to make a context, it fails with `UnknownError("glxSwapIntervalMESA failed")`. I've included at the end a backtrace running the `hello_world` example (as of commit `40d70a79bb3f46c75b4d11bd4424c036c540a9fd`).

This is running in a VM, which seems extremely likely to be the issue at some level. It's a Ubuntu 16.04 VM running on a Windows 7 host in VMWare Workstation 12. Here's the output of [glxinfo](https://github.com/ggez/ggez/files/1761581/glxinfo.txt).

I've tried modify the OpenGL version used, to no avail: I've tried `3.0`, `2.1`, and `1.1`. Here's the diff showing what I did in case I did it wrong:
```
diff --git a/examples/hello_world.rs b/examples/hello_world.rs
index 58e8e9b..c3138fd 100644
--- a/examples/hello_world.rs
+++ b/examples/hello_world.rs
@@ -62,7 +62,8 @@ impl event::EventHandler for MainState {
 // do the work of creating our MainState and running our game,
 // * then just call `game.run()` which runs the `Game` mainloop.
 pub fn main() {
-    let c = conf::Conf::new();
+    let mut c = conf::Conf::new();
+    c.backend = conf::Backend::OpenGL{major: 3, minor:0};
     let ctx = &mut Context::load_from_conf("helloworld", "ggez", c).unwrap();
 
     // We add the CARGO_MANIFEST_DIR/resources do the filesystems paths so
```
And here's the backtrace:
```
cargo run --example hello_world
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `target/debug/examples/hello_world`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: UnknownError("glxSwapIntervalMESA failed")', libcore/result.rs:945:5
stack backtrace:
   0:     0x55ef91ea6c63 - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::h038efa1ead8ee277
                               at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1:     0x55ef91ea14c4 - std::sys_common::backtrace::_print::haa2469a6e6008f11
                               at libstd/sys_common/backtrace.rs:71
   2:     0x55ef91ea5c0d - std::panicking::default_hook::{{closure}}::ha5918fb5e6555fa7
                               at libstd/sys_common/backtrace.rs:59
                               at libstd/panicking.rs:206
   3:     0x55ef91ea5909 - std::panicking::default_hook::h4ddc2e74f1ee24ea
                               at libstd/panicking.rs:222
   4:     0x55ef91ea606f - std::panicking::rust_panic_with_hook::hf7d91abe57c1b397
                               at libstd/panicking.rs:400
   5:     0x55ef91ea5e92 - std::panicking::begin_panic_fmt::h83678fb4d002c950
                               at libstd/panicking.rs:347
   6:     0x55ef91ea5dc2 - rust_begin_unwind
                               at libstd/panicking.rs:323
   7:     0x55ef91eec670 - core::panicking::panic_fmt::h28698393dd7f7d54
                               at libcore/panicking.rs:71
   8:     0x55ef91986657 - core::result::unwrap_failed::hb6a6cad29b2abe59
                               at /checkout/src/libcore/macros.rs:23
   9:     0x55ef919864b6 - <core::result::Result<T, E>>::unwrap::h87e337a4feea38b9
                               at /checkout/src/libcore/result.rs:782
  10:     0x55ef91991f0d - hello_world::main::h2b713c399af204e2
                               at examples/hello_world.rs:66
  11:     0x55ef91992f01 - std::rt::lang_start::{{closure}}::h4119ece9ec322894
                               at /checkout/src/libstd/rt.rs:74
  12:     0x55ef91ea5d27 - std::panicking::try::do_call::hc331ed8d98dd7e72
                               at libstd/rt.rs:59
                               at libstd/panicking.rs:305
  13:     0x55ef91eb1f9e - __rust_maybe_catch_panic
                               at libpanic_unwind/lib.rs:102
  14:     0x55ef91ea6533 - std::rt::lang_start_internal::h2271475d0185b808
                               at libstd/panicking.rs:284
                               at libstd/panic.rs:361
                               at libstd/rt.rs:58
  15:     0x55ef91992ee1 - std::rt::lang_start::h376a47b757c20492
                               at /checkout/src/libstd/rt.rs:74
  16:     0x55ef919923fd - main
  17:     0x7fa35d9b482f - __libc_start_main
  18:     0x55ef91985858 - _start
  19:                0x0 - <unknown>
```

**Comment by icefoxen** [comment]
Ooh, that's a new one.

It looks like `glxSwapInterval` is how one sets vsync.  Try setting `conf.window_mode.vsync` to false and see if that changes anything; it's true by default.

If it still fails, you may wish to try modifying the source to remove [this call](https://github.com/ggez/ggez/blob/master/src/graphics/context.rs#L97) and see if that changes anything.

**Comment by rocurley** [comment]
Ok! Setting vsync to false didn't do it, but if I comment out `set_vsync` I get a different error: `UnknownError("")`. It comes from the call to `display_dpi`. If I give it a default value, I'm able to run the hello world example. The diff that gets it working for me:
```
diff --git a/src/graphics/context.rs b/src/graphics/context.rs
index 5ef802d..10aed8b 100644
--- a/src/graphics/context.rs
+++ b/src/graphics/context.rs
@@ -94,10 +94,11 @@ impl GraphicsContext {
         let (window, gl_context, device, mut factory, screen_render_target, depth_view) =
             gfx_window_sdl::init(window_builder)?;
 
-        GraphicsContext::set_vsync(video, window_mode.vsync);
+        //GraphicsContext::set_vsync(video, window_mode.vsync);
 
         let display_index = window.display_index()?;
-        let dpi = window.subsystem().display_dpi(display_index)?;
+        let dpi_result = window.subsystem().display_dpi(display_index);
+        let dpi = dpi_result.unwrap_or((90.,90.,90./2.0_f32.sqrt()));
 
         // GFX SETUP
         let mut encoder: gfx::Encoder<
```

**Comment by icefoxen** [comment]
Glad it works!  But man that's wacky.  I especially like the empty string as the error message.

These are both probably bugs either in SDL, the video driver, or both.  Would be nice to see if we can reproduce just using raw SDL, either in Rust or C.

**Comment by icefoxen** [comment]
Hmm, assuming you're using x11 instead of wayland, the `display_dpi()` call should be ending up [here](https://hg.libsdl.org/SDL/file/2088cd828335/src/video/x11/SDL_x11modes.c#l1059).  But there seems no way that could produce an empty string, and the [rust wrapper](https://github.com/Rust-SDL2/rust-sdl2/blob/329334c8a6e43bac615ad747fa2e65ab7e0cf56a/src/sdl2/sdl.rs#L314) doesn't either.

Weeeeeeird.

**Comment by rocurley** [comment]
I think that the empty string error message is fixed by [this patch](https://hg.libsdl.org/SDL/rev/b4316ed0c72c). My version of SDL (2.0.4) predates that.

**Comment by icefoxen** [comment]
Aha!  That would do it!  Would be nice to have a better way of setting DPI in ggez though, beyond the option of just "set it by hand".  It is *annoyingly* troublesome.  Not sure how to do this well though.

**Comment by rocurley** [comment]
I didn't notice this the first time, but `set_vsync` doesn't actually return anything. Indeed, having "fixed" the DPI issue, I can uncomment the vsync line fine. AFAICT, something (recoverable) goes wrong in SDL relating to `set_vsync`, setting the error message I originally reported, and the missing error message in `display_dpi()` means that the error from vsync is returned instead.

**Comment by icefoxen** [comment]
Oooooh.  The sdl2 crate's `gl_set_swap_interval` returns a bool, not a result, and ggez doesn't check that bool.  Good catch!  Pushed that one upstream here: https://github.com/Rust-SDL2/rust-sdl2/issues/755 , but in the mean time it's an easy fix in ggez to just check the result of `gl_set_swap_interval` properly and get the error message if it fails.  ...though now I have to come up with a way of telling the user about a soft error.

**Comment by icefoxen** [comment]
Should be fixed-ish in the latest commit; it is still a hard error (it won't create a `Context` if it can't set the vsync appropriately), but it should at least give you a sensible error message.

Might end up making an option to specify whether "can't set vsync" is a hard or soft error?  Or having it always be soft but giving you a way to check for it?  I dunno.

**Comment by rocurley** [comment]
Thanks!

This does seems like a situation where you'd like to make as many errors soft as possible: it's not like the people using ggez are going to repent and write better code if the errors are hard, since the problematic code is presumably their drivers. I'd even favor making the DPI check recoverable (of course, I'm motivated by the prospect of being able to use ggez without maintaining the world's tiniest fork).

All that said, I'm not sure what a good way to deal with soft errors would be: maybe also return a vector of errors?

**Comment by jdeeny** [comment]
I am trying to use ggez on linux under vmware with a windows host and running into the DPI error. If I add a default like @rocurley did above, it works fine.

I don't have the issue on my linux desktop, or my friends windows desktops. And that is the bigger problem, I think. If the error isn't fatal and has a sane default that won't harm the user experience too badly, ggez should warn the dev (could be something like an error vec they could log) and keep on trucking. Otherwise you end up with preventable user experience problems in situations the dev may not (or can not) know about in advance.

I'd much rather a user have to complain about/bug report "DPI messed up on vmware" over "game crashes immediately and says Unknown Error: No DPI".

**Comment by icefoxen** [comment]
Definitely.  Just need to think of a way to communicate soft errors nicely.  :-)

Options: 

 * The Easy way is to have the programmer set the setting, then get it right afterwards to see if what it actually is matches with what they set.  Advantages: It works without any API alterations or needing to keep track of anything new.  Disadvantages: It fails silently, it's up to the programmer to check for the error, which (to a first approximation) nobody will.
 * Use a logging framework, like slog or such.  Pros: Easy, clear, easy to integrate with other programs.  Cons: Messy, easily becomes a garbage dump of information, imposes more policy on the programmer (Thou Shalt Use Slog For Logging or whatever).  Careful setup of the logging structure could avoid this, but you end up with something that's mostly human-readable instead of machine-readable still.
 * Record warnings somehow and let the programmer retrieve them; just have a `Vec<WarningType>` or whatever.  Pros: Simple, well-defined, easy.  Cons: Basically becomes a custom logging framework very quickly, with all the pro's and cons of that.
 * Somehow make warnings part of the API, similar to `Result`?  I like this best but I'm not sure how to do it yet.  Perhaps something like `Result<Option<Warning>, Error>`?  That might get annoying when the return type becomes `Result<(T, Option<Warning>), Error>`, but a little bit of clever design should make that able to be handled with combinators.

**Comment by Ratysz** [comment]
> Use a logging framework, like slog or such.

I'm currently getting a lot of mileage from [log](https://crates.io/crates/log) crate, via [simplelog](https://crates.io/crates/simplelog). As far as I can tell, it's entirely opt-in for executables, even if dependencies (read, `ggez`) talk with `log`.

**Comment by jdeeny** [comment]
Could the existing data structure be changed to contain the state of the values without breaking compatibility? Could an enum with `Some(`, `SomeDefault`, `SomeDefaultWithWarning`, `NoneWithRecoverableError`, `NoneWithFatalError` be used? (maybe with overloading to maintain the existing api for get/set, my rust is fuzzy here) Some iterators could be provided to let the user scan the list for problems and log/correct as appropriate. Attempting an operation that is blocked by an error can point to the source of the problem.

**Comment by icefoxen** [comment]
I've changed my opinion on this since there will always be errors and I'm not going to try to figure out which ones are okay to power through and which ones should be actually given to the developer. It's not ggez's job to make those decisions. The point of returning a distinct error value instead of panicking is to let the developer see what went wrong and know how to retry it if possible. In this case, it really should be possible to query the graphics driver to see if it supports vsync or dpi querying, but I don't think that's currently possible through SDL.

Saying things like "well you asked for X and I can't do it so I'll just do Y instead" isn't really good software practice; I'll leave that up to the opengl drivers.
